package drizzle;

message Table {
  required string name = 1;
  required string engine = 2;

  enum FieldType {
    DOUBLE = 0;
    VARCHAR = 1;
    TEXT = 2;
    BLOB = 3;
    ENUM = 4;
    SET = 5;
    TINYINT = 6;
    SMALLINT = 7;
    INTEGER = 8;
    BIGINT = 9;
    DECIMAL = 10;
    VARBINARY = 11;
    DATE = 12;
    TIME = 13;
    TIMESTAMP = 14;
    DATETIME = 15;
  }

  message Field {
    required string name = 1;
    required FieldType type = 2;
    optional string collation = 3;
    optional string comment = 4;
    optional bool unique = 5;
    optional bool autoincrement = 6;
    optional bool key = 7;
    optional bool primary = 8;
    optional bool is_unsigned = 9;
    optional string custom_name = 10;
    optional bool is_notnull = 11 [default = false];
    optional int32 scale = 12;
    optional string characterset = 13;
    optional int32 length = 14;
    optional string default_value = 15;
    repeated string values = 16;
    optional bool on_update = 17;
  }

  message KeyPart {
    required string name = 1;
    optional int32 length = 2;
  }

  enum IndexType {
    ORDERED = 0;
    HASH = 1;
  }

  message Index {
    required string name = 1;
    repeated KeyPart values = 2;
    optional bool primary = 4;
    optional int32 key_block_size = 5;
    optional IndexType type = 6;
  }

  repeated Field field = 4;
  repeated Index index = 5;

  optional string primary = 6;
  optional int32  auto_increment = 7;
  optional int32 avg_row_length = 8;
  optional string character_set = 9;
  optional bool checksum = 10;
  optional string collation = 11;
  optional string comment = 12;
  optional string connection = 13;
  optional string data_directory = 14;
  optional string index_directory = 15;
  optional bool delay_key_write = 16;
  optional int32 max_rows = 17;
  optional int32 min_rows = 18;
  optional bool pack_keys = 19;
  optional string row_format = 20;
  optional string characterset = 21;
  optional bool temp = 22;
}

message TableList {
  repeated Table table = 1;
}
